#### <u>Business Understanding</u>:
In the telecom industry, customers are able to choose from multiple service providers and actively switch from 
one operator to another. In this highly competitive market, the telecommunications industry experiences an 
average of 15-25% annual churn rate. Given the fact that it costs 5-10 times more to acquire a new customer 
than to retain an existing one, customer retention has now become even more important than customer acquisition.
 
For many operators, retaining high profitable customers is the number one business goal.
  
<b>Strategy:</b> Based on the existing customer datasets, to predict which customers are at high risk of churn and 
recommend strategies to manage customer churn based on the observations.

<b>Business Objective:</b> The objective is to predict the churn in the last (i.e. the ninth) month using the data (features) from the first three months. To do this task well, understanding the typical customer behaviour during churn will be helpful.

<b><u>Customer Behaviour During Churn</u></b>

In the given dataset of four-month window;

   <b>Good Phase  :</b> First Two Months<br>
   <b>Action Phase:</b> Third Month<br>
   <b>Churn Phase :</b> Fourth Month

<b><u>Preprocessing Rules</u></b>

   1. Consider only <b>High-value customers:</b> approximately 80% of revenue comes from the top 20% customers (called high-value customers). Thus, if we can reduce churn of the high-value customers, we will be able to reduce significant revenue leakage.
   2. Consider Only <b>Usage-based churn:</b> Customers who have not done any usage, either incoming or outgoing - in terms of calls, internet etc. over a period of time.
   
<b><u>Data Preparation:</u></b>

The following data preparation steps are crucial for this problem:
 
   1. Derive new features 
   2. Filter high-value customers 
       - Based 70th percentile of the average recharge amount in the first two months.
   3. Tag churners and remove attributes of the churn phase
       - Attributes (total_ic_mou_9, total_og_mou_9, vol_2g_mb_9, vol_3g_mb_9).

<b>Note:</b> After tagging churners, Remove all the attributes corresponding to the churn phase (Name starts with'_9').


<b><u>Suggested Steps:</u></b>

   1. Preprocess data (convert columns to appropriate formats, handle missing values, etc.).
   2. Conduct appropriate exploratory analysis to extract useful insights.
   3. Derive new features.
   4. Reduce the number of variables using PCA.
   5. Train a variety of models, tune model hyperparameters, etc. (handle class imbalance).
   6. Evaluate the models using appropriate evaluation metrics. 
   7. choose a model based on some evaluation metric.
   8. Identifying important predictor attributes which help the business to understand indicators of churn.
   9. After identifying important predictors, display them visually.<br>
   10. Recommend strategies to manage customer churn based on your observations.

#### <u>Data Sources</u>:

1. DataDictionaryTelecomChurnCaseStudy.xlsx --> Provides meaning of the column attributes for the given customer dataset.
2. telecom_churn_data.csv              --> Customer data for four months between Jun-2014 to Sep-2014.

# Reading and Understanding the Data

In [9]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.decomposition import PCA
from sklearn.decomposition import IncrementalPCA

from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestClassifier

from sklearn.model_selection import KFold, cross_val_score, train_test_split
from sklearn.metrics import mean_squared_error, r2_score

import xgboost as xgb
from xgboost import XGBClassifier

from sklearn import metrics
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score, roc_curve, auc

import warnings
warnings.filterwarnings('ignore')

pd.set_option("display.max_columns", 500)
pd.set_option("display.max_rows", 1000)

In [5]:
data = pd.read_csv("telecom_churn_data.csv")
print(data.shape)
data.sample(5)

(99999, 226)


mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  \
37480     7000417366        109             0.0             0.0   
8066      7000651270        109             0.0             0.0   
92568     7001597106        109             0.0             0.0   
31127     7001517345        109             0.0             0.0   
62924     7001464712        109             0.0             0.0   

       loc_ic_t2o_mou last_date_of_month_6 last_date_of_month_7  \
37480             0.0            6/30/2014            7/31/2014   
8066              0.0            6/30/2014            7/31/2014   
92568             0.0            6/30/2014            7/31/2014   
31127             0.0            6/30/2014            7/31/2014   
62924             0.0            6/30/2014            7/31/2014   

      last_date_of_month_8 last_date_of_month_9   arpu_6  ...  sachet_3g_9  \
37480            8/31/2014            9/30/2014  105.692  ...            0   
8066             8/31/2014            9/30/2014  122.782  ...            0   
92568            8/31/2014            9/30/2014   41.130  ...            0   
31127            8/31/2014            9/30/2014  377.309  ...            0   
62924            8/31/2014            9/30/2014  192.035  ...            0   

       fb_user_6  fb_user_7  fb_user_8  fb_user_9   aon  aug_vbc_3g  \
37480        NaN        NaN        NaN        NaN   375        0.00   
8066         1.0        1.0        1.0        1.0   304      579.74   
92568        NaN        NaN        NaN        NaN  1679        0.00   
31127        NaN        NaN        0.0        0.0   999        0.00   
62924        NaN        NaN        NaN        NaN  2837        0.00   

       jul_vbc_3g  jun_vbc_3g  sep_vbc_3g  
37480        0.00        0.00         0.0  
8066       934.66      506.33         0.0  
92568        0.00        0.00         0.0  
31127        0.00        0.00         0.0  
62924        0.00        0.00         0.0  

[5 rows x 226 columns]

In [6]:
# Checking the data insight
data.describe()

mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  \
count   9.999900e+04    99999.0         98981.0         98981.0   
mean    7.001207e+09      109.0             0.0             0.0   
std     6.956694e+05        0.0             0.0             0.0   
min     7.000000e+09      109.0             0.0             0.0   
25%     7.000606e+09      109.0             0.0             0.0   
50%     7.001205e+09      109.0             0.0             0.0   
75%     7.001812e+09      109.0             0.0             0.0   
max     7.002411e+09      109.0             0.0             0.0   

       loc_ic_t2o_mou        arpu_6        arpu_7        arpu_8        arpu_9  \
count         98981.0  99999.000000  99999.000000  99999.000000  99999.000000   
mean              0.0    282.987358    278.536648    279.154731    261.645069   
std               0.0    328.439770    338.156291    344.474791    341.998630   
min               0.0  -2258.709000  -2014.045000   -945.808000  -1899.505000   
25%               0.0     93.411500     86.980500     84.126000     62.685000   
50%               0.0    197.704000    191.640000    192.080000    176.849000   
75%               0.0    371.060000    365.344500    369.370500    353.466500   
max               0.0  27731.088000  35145.834000  33543.624000  38805.617000   

        onnet_mou_6  ...   sachet_3g_9     fb_user_6     fb_user_7  \
count  96062.000000  ...  99999.000000  25153.000000  25571.000000   
mean     132.395875  ...      0.084581      0.914404      0.908764   
std      297.207406  ...      0.650457      0.279772      0.287950   
min        0.000000  ...      0.000000      0.000000      0.000000   
25%        7.380000  ...      0.000000      1.000000      1.000000   
50%       34.310000  ...      0.000000      1.000000      1.000000   
75%      118.740000  ...      0.000000      1.000000      1.000000   
max     7376.710000  ...     49.000000      1.000000      1.000000   

          fb_user_8     fb_user_9           aon    aug_vbc_3g    jul_vbc_3g  \
count  26339.000000  25922.000000  99999.000000  99999.000000  99999.000000   
mean       0.890808      0.860968   1219.854749     68.170248     66.839062   
std        0.311885      0.345987    954.733842    267.580450    271.201856   
min        0.000000      0.000000    180.000000      0.000000      0.000000   
25%        1.000000      1.000000    467.000000      0.000000      0.000000   
50%        1.000000      1.000000    863.000000      0.000000      0.000000   
75%        1.000000      1.000000   1807.500000      0.000000      0.000000   
max        1.000000      1.000000   4337.000000  12916.220000   9165.600000   

         jun_vbc_3g    sep_vbc_3g  
count  99999.000000  99999.000000  
mean      60.021204      3.299373  
std      253.938223     32.408353  
min        0.000000      0.000000  
25%        0.000000      0.000000  
50%        0.000000      0.000000  
75%        0.000000      0.000000  
max    11166.210000   2618.570000  

[8 rows x 214 columns]

In [7]:
# Checking for dupicates in mobile number column
sum(data.duplicated(subset= "mobile_number"))

0

**Initial Observations**
- There are 226 columns mostly divided for 4 months 6-9
- To decide the churn rate columns related to revenue, recharge, usage are important
- As per given information we need to identify the prime customers which needs analysis of behaviour of customer
- As every feature has total column also for a particular month, it is useful for analysis

In [10]:
# Checking Null Values
data.isnull().sum()

mobile_number                   0
circle_id                       0
loc_og_t2o_mou               1018
std_og_t2o_mou               1018
loc_ic_t2o_mou               1018
last_date_of_month_6            0
last_date_of_month_7          601
last_date_of_month_8         1100
last_date_of_month_9         1659
arpu_6                          0
arpu_7                          0
arpu_8                          0
arpu_9                          0
onnet_mou_6                  3937
onnet_mou_7                  3859
onnet_mou_8                  5378
onnet_mou_9                  7745
offnet_mou_6                 3937
offnet_mou_7                 3859
offnet_mou_8                 5378
offnet_mou_9                 7745
roam_ic_mou_6                3937
roam_ic_mou_7                3859
roam_ic_mou_8                5378
roam_ic_mou_9                7745
roam_og_mou_6                3937
roam_og_mou_7                3859
roam_og_mou_8                5378
roam_og_mou_9                7745
loc_og_t2t_mou

In [11]:
# Extracting columns which represent some total
monthwise_total = data.columns[data.columns.str.contains('total')]
monthwise_total

Index(['total_og_mou_6', 'total_og_mou_7', 'total_og_mou_8', 'total_og_mou_9',
       'total_ic_mou_6', 'total_ic_mou_7', 'total_ic_mou_8', 'total_ic_mou_9',
       'total_rech_num_6', 'total_rech_num_7', 'total_rech_num_8',
       'total_rech_num_9', 'total_rech_amt_6', 'total_rech_amt_7',
       'total_rech_amt_8', 'total_rech_amt_9', 'total_rech_data_6',
       'total_rech_data_7', 'total_rech_data_8', 'total_rech_data_9'],
      dtype='object')

In [12]:
# Extracting columns which represent some recharge
monthwise_recharges = data.columns[data.columns.str.contains('rech_amt')]
monthwise_recharges

Index(['total_rech_amt_6', 'total_rech_amt_7', 'total_rech_amt_8',
       'total_rech_amt_9', 'max_rech_amt_6', 'max_rech_amt_7',
       'max_rech_amt_8', 'max_rech_amt_9', 'av_rech_amt_data_6',
       'av_rech_amt_data_7', 'av_rech_amt_data_8', 'av_rech_amt_data_9'],
      dtype='object')

## Churn or Not Churn

**As per the given context:**
- We need to create new column for labeling churn or not churn as per the 9th month usage information
- After that need to drop the 9th months data

In [16]:
# Finding all columns related to september month
sep_col = [x for x in list(data) if "_9" in x or "sep" in x or "Sep" in x or "September" in x or "september" in x]
print(sep_col)


#from these columns to detemine churn or not we only need information from incoming, outgoing and data usage.
data['churn'] = np.where(data[['vol_2g_mb_9','vol_3g_mb_9', 'total_ic_mou_9','total_og_mou_9']].sum(axis = 1) == 0,1,0)
print(data.shape)
data.head(3)

['last_date_of_month_9', 'arpu_9', 'onnet_mou_9', 'offnet_mou_9', 'roam_ic_mou_9', 'roam_og_mou_9', 'loc_og_t2t_mou_9', 'loc_og_t2m_mou_9', 'loc_og_t2f_mou_9', 'loc_og_t2c_mou_9', 'loc_og_mou_9', 'std_og_t2t_mou_9', 'std_og_t2m_mou_9', 'std_og_t2f_mou_9', 'std_og_t2c_mou_9', 'std_og_mou_9', 'isd_og_mou_9', 'spl_og_mou_9', 'og_others_9', 'total_og_mou_9', 'loc_ic_t2t_mou_9', 'loc_ic_t2m_mou_9', 'loc_ic_t2f_mou_9', 'loc_ic_mou_9', 'std_ic_t2t_mou_9', 'std_ic_t2m_mou_9', 'std_ic_t2f_mou_9', 'std_ic_t2o_mou_9', 'std_ic_mou_9', 'total_ic_mou_9', 'spl_ic_mou_9', 'isd_ic_mou_9', 'ic_others_9', 'total_rech_num_9', 'total_rech_amt_9', 'max_rech_amt_9', 'date_of_last_rech_9', 'last_day_rch_amt_9', 'date_of_last_rech_data_9', 'total_rech_data_9', 'max_rech_data_9', 'count_rech_2g_9', 'count_rech_3g_9', 'av_rech_amt_data_9', 'vol_2g_mb_9', 'vol_3g_mb_9', 'arpu_3g_9', 'arpu_2g_9', 'night_pck_user_9', 'monthly_2g_9', 'sachet_2g_9', 'monthly_3g_9', 'sachet_3g_9', 'fb_user_9', 'sep_vbc_3g']
(99999, 22

mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou  \
0     7000842753        109             0.0             0.0             0.0   
1     7001865778        109             0.0             0.0             0.0   
2     7001625959        109             0.0             0.0             0.0   

  last_date_of_month_6 last_date_of_month_7 last_date_of_month_8  \
0            6/30/2014            7/31/2014            8/31/2014   
1            6/30/2014            7/31/2014            8/31/2014   
2            6/30/2014            7/31/2014            8/31/2014   

  last_date_of_month_9   arpu_6   arpu_7   arpu_8   arpu_9  onnet_mou_6  \
0            9/30/2014  197.385  214.816  213.803   21.100          NaN   
1            9/30/2014   34.047  355.074  268.321   86.285        24.11   
2            9/30/2014  167.690  189.058  210.226  290.714        11.54   

   onnet_mou_7  onnet_mou_8  onnet_mou_9  offnet_mou_6  offnet_mou_7  \
0          NaN         0.00          NaN           NaN           NaN   
1        78.68         7.68        18.34         15.74         99.84   
2        55.24        37.26        74.81        143.33        220.59   

   offnet_mou_8  offnet_mou_9  roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  \
0          0.00           NaN            NaN            NaN            0.0   
1        304.76         53.76            0.0            0.0            0.0   
2        208.36        118.91            0.0            0.0            0.0   

   roam_ic_mou_9  roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  roam_og_mou_9  \
0            NaN            NaN            NaN            0.0            NaN   
1           0.00            0.0            0.0            0.0           0.00   
2          38.49            0.0            0.0            0.0          70.94   

   loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  loc_og_t2t_mou_9  \
0               NaN               NaN              0.00               NaN   
1             23.88             74.56              7.68             18.34   
2              7.19             28.74             13.58             14.39   

   loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  loc_og_t2m_mou_9  \
0               NaN               NaN              0.00               NaN   
1             11.51             75.94            291.86             53.76   
2             29.34             16.86             38.46             28.16   

   loc_og_t2f_mou_6  loc_og_t2f_mou_7  loc_og_t2f_mou_8  loc_og_t2f_mou_9  \
0               NaN               NaN              0.00               NaN   
1              0.00              0.00              0.00              0.00   
2             24.11             21.79             15.61             22.24   

   loc_og_t2c_mou_6  loc_og_t2c_mou_7  loc_og_t2c_mou_8  loc_og_t2c_mou_9  \
0               NaN               NaN              0.00               NaN   
1               0.0              2.91              0.00              0.00   
2               0.0            135.54             45.76              0.48   

   loc_og_mou_6  loc_og_mou_7  loc_og_mou_8  loc_og_mou_9  std_og_t2t_mou_6  \
0           NaN           NaN          0.00           NaN               NaN   
1         35.39        150.51        299.54         72.11              0.23   
2         60.66         67.41         67.66         64.81              4.34   

   std_og_t2t_mou_7  std_og_t2t_mou_8  std_og_t2t_mou_9  std_og_t2m_mou_6  \
0               NaN              0.00               NaN               NaN   
1              4.11              0.00              0.00              0.00   
2             26.49             22.58              8.76             41.81   

   std_og_t2m_mou_7  std_og_t2m_mou_8  std_og_t2m_mou_9  std_og_t2f_mou_6  \
0               NaN              0.00               NaN               NaN   
1              0.46              0.13              0.00              0.00   
2             67.41             75.53              9.28              1.48   

   std_og_t2f_mou_

In [18]:
#Checking data balance for Churning 
print(data.churn.value_counts()/len(data))


0    0.898089
1    0.101911
Name: churn, dtype: float64


In [19]:
# Dropping information related to column 9 as it will not be available for prediction
data = data.drop(columns=sep_col, axis = 1)
print(data.shape)

(99999, 172)


**As per the problem statement the high-value customers shoud be defined as:**

Those who have recharged with an amount more than or equal to X, where X is the 70th percentile of the average recharge amount in the first two months (the good phase).

- To identify the high value customers we need to check the total usage 
- We have 2 kinds of recharge 1) cellular 2) Data
- We will find the total recharge amount for month 6 & 7 and create a new dataframe for prime(high value) customers only

In [20]:
# Extracting columns which represent some recharge
monthwise_recharges = data.columns[data.columns.str.contains('rech_amt')]
monthwise_recharges

monthwise_data_recharges_cols = data.columns[data.columns.str.contains('_data_')]
monthwise_data_recharges_cols
# from above column list we have total recharged data but not the cost; we can get it by multiplying the avg data rech amnt

Index(['date_of_last_rech_data_6', 'date_of_last_rech_data_7',
       'date_of_last_rech_data_8', 'total_rech_data_6', 'total_rech_data_7',
       'total_rech_data_8', 'max_rech_data_6', 'max_rech_data_7',
       'max_rech_data_8', 'av_rech_amt_data_6', 'av_rech_amt_data_7',
       'av_rech_amt_data_8'],
      dtype='object')

**Derived few columns as per requirement**

In [22]:
# derived Columns which are required for further calculation
columns_to_consider_cell = ['total_rech_amt_6', 'total_rech_amt_7', 'total_rech_amt_8']
columns_to_consider_data = ['total_rech_data_6', 'total_rech_data_7', 'total_rech_data_8',
                      'av_rech_amt_data_6', 'av_rech_amt_data_7', 'av_rech_amt_data_8']

In [23]:
# Calculate the total data recharge amount 
data['total_data_rech_amt_6'] = data['total_rech_data_6'] * data['av_rech_amt_data_6'] 
data['total_data_rech_amt_7'] = data['total_rech_data_7'] * data['av_rech_amt_data_7']
data['total_data_rech_amt_8'] = data['total_rech_data_8'] * data['av_rech_amt_data_8']
data.shape

(99999, 175)

In [24]:
# now we don't need other 2 columns as we have calculated the total data recharge amount hence dropping them 
data = data.drop(columns=columns_to_consider_data, axis = 1)
data.shape

(99999, 169)

In [25]:
# To calculate total recharge amount we need following columns
total_rech_6_and_7_cols = ['total_data_rech_amt_6', 'total_data_rech_amt_7', 'total_rech_amt_6', 'total_rech_amt_7']

# Check if null values are there
print(data[total_rech_6_and_7_cols].isnull().sum())

# We will replace null with 0 for now
data[total_rech_6_and_7_cols] = data[total_rech_6_and_7_cols].fillna(0)

# Check if null values are there
print(data[total_rech_6_and_7_cols].isnull().sum())

total_data_rech_amt_6    74846
total_data_rech_amt_7    74428
total_rech_amt_6             0
total_rech_amt_7             0
dtype: int64
total_data_rech_amt_6    0
total_data_rech_amt_7    0
total_rech_amt_6         0
total_rech_amt_7         0
dtype: int64


In [26]:
# Now we will derive new column as total_avg_rech_amt_6_and_7 to store total avg recharge amount for month 6 and 7
total_avg_rech_amt_6_and_7 = (data['total_data_rech_amt_6'] + data['total_data_rech_amt_7'] \
                                 + data['total_rech_amt_6'] + data['total_rech_amt_7'])/2
print("Total Avg amount of recharge for 6 and 7 mnth is: ", sum(total_avg_rech_amt_6_and_7) , "\n\n")
data.head(5)

Total Avg amount of recharge for 6 and 7 mnth is:  48532321.43 




mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou  \
0     7000842753        109             0.0             0.0             0.0   
1     7001865778        109             0.0             0.0             0.0   
2     7001625959        109             0.0             0.0             0.0   
3     7001204172        109             0.0             0.0             0.0   
4     7000142493        109             0.0             0.0             0.0   

  last_date_of_month_6 last_date_of_month_7 last_date_of_month_8   arpu_6  \
0            6/30/2014            7/31/2014            8/31/2014  197.385   
1            6/30/2014            7/31/2014            8/31/2014   34.047   
2            6/30/2014            7/31/2014            8/31/2014  167.690   
3            6/30/2014            7/31/2014            8/31/2014  221.338   
4            6/30/2014            7/31/2014            8/31/2014  261.636   

    arpu_7   arpu_8  onnet_mou_6  onnet_mou_7  onnet_mou_8  offnet_mou_6  \
0  214.816  213.803          NaN          NaN         0.00           NaN   
1  355.074  268.321        24.11        78.68         7.68         15.74   
2  189.058  210.226        11.54        55.24        37.26        143.33   
3  251.102  508.054        99.91        54.39       310.98        123.31   
4  309.876  238.174        50.31       149.44        83.89         76.96   

   offnet_mou_7  offnet_mou_8  roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  \
0           NaN          0.00            NaN            NaN           0.00   
1         99.84        304.76            0.0           0.00           0.00   
2        220.59        208.36            0.0           0.00           0.00   
3        109.01         71.68            0.0          54.86          44.38   
4         91.88        124.26            0.0           0.00           0.00   

   roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  loc_og_t2t_mou_6  \
0            NaN            NaN           0.00               NaN   
1            0.0           0.00           0.00             23.88   
2            0.0           0.00           0.00              7.19   
3            0.0          28.09          39.04             73.68   
4            0.0           0.00           0.00             50.31   

   loc_og_t2t_mou_7  loc_og_t2t_mou_8  loc_og_t2m_mou_6  loc_og_t2m_mou_7  \
0               NaN              0.00               NaN               NaN   
1             74.56              7.68             11.51             75.94   
2             28.74             13.58             29.34             16.86   
3             34.81             10.61            107.43             83.21   
4            149.44             83.89             67.64             91.88   

   loc_og_t2m_mou_8  loc_og_t2f_mou_6  loc_og_t2f_mou_7  loc_og_t2f_mou_8  \
0              0.00               NaN               NaN              0.00   
1            291.86              0.00              0.00              0.00   
2             38.46             24.11             21.79             15.61   
3             22.46              1.91              0.65              4.91   
4            124.26              0.00              0.00              0.00   

   loc_og_t2c_mou_6  loc_og_t2c_mou_7  loc_og_t2c_mou_8  loc_og_mou_6  \
0               NaN               NaN              0.00           NaN   
1               0.0              2.91              0.00         35.39   
2               0.0            135.54             45.76         60.66   
3               0.0              0.00              0.00        183.03   
4               0.0              0.00              0.00        117.96   

   loc_og_mou_7  loc_og_mou_8  std_og_t2t_mou_6  std_og_t2t_mou_7  \
0           NaN          0.00               NaN               NaN   
1        150.51        299.54              0.23              4.11   
2         67.41         67.66              4.34             26.49   
3        118.68         37.99             26.23             14.89   
4        241.33     

In [27]:
# We need highvalue customers which have 70% or more avg recharge amount (6 & 7 month)
prime_cust_threshold = np.percentile(total_avg_rech_amt_6_and_7, 70.0)
print("The prime customer threshold recharge amount is: ", prime_cust_threshold)

The prime customer threshold recharge amount is:  478.0


In [28]:
# Now we will create new dataframe which will have only prime customer with percentile recharge >=70
prime = data[total_avg_rech_amt_6_and_7 >= prime_cust_threshold]
prime.shape

(30001, 169)

In [30]:
# Checking the data balance
print(prime.churn.value_counts()/len(prime))
prime_backup = prime.copy()

prime.head(3)

0    0.918636
1    0.081364
Name: churn, dtype: float64


mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou  \
0     7000842753        109             0.0             0.0             0.0   
7     7000701601        109             0.0             0.0             0.0   
8     7001524846        109             0.0             0.0             0.0   

  last_date_of_month_6 last_date_of_month_7 last_date_of_month_8    arpu_6  \
0            6/30/2014            7/31/2014            8/31/2014   197.385   
7            6/30/2014            7/31/2014            8/31/2014  1069.180   
8            6/30/2014            7/31/2014            8/31/2014   378.721   

     arpu_7    arpu_8  onnet_mou_6  onnet_mou_7  onnet_mou_8  offnet_mou_6  \
0   214.816   213.803          NaN          NaN         0.00           NaN   
7  1349.850  3171.480        57.84        54.68        52.29        453.43   
8   492.223   137.362       413.69       351.03        35.08         94.66   

   offnet_mou_7  offnet_mou_8  roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  \
0           NaN          0.00            NaN            NaN           0.00   
7        567.16        325.91          16.23          33.49          31.64   
8         80.63        136.48           0.00           0.00           0.00   

   roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  loc_og_t2t_mou_6  \
0            NaN            NaN           0.00               NaN   
7          23.74          12.59          38.06             51.39   
8           0.00           0.00           0.00            297.13   

   loc_og_t2t_mou_7  loc_og_t2t_mou_8  loc_og_t2m_mou_6  loc_og_t2m_mou_7  \
0               NaN              0.00               NaN               NaN   
7             31.38             40.28            308.63            447.38   
8            217.59             12.49             80.96             70.58   

   loc_og_t2m_mou_8  loc_og_t2f_mou_6  loc_og_t2f_mou_7  loc_og_t2f_mou_8  \
0              0.00               NaN               NaN              0.00   
7            162.28             62.13             55.14             53.23   
8             50.54              0.00              0.00              0.00   

   loc_og_t2c_mou_6  loc_og_t2c_mou_7  loc_og_t2c_mou_8  loc_og_mou_6  \
0               NaN               NaN              0.00           NaN   
7               0.0               0.0              0.00        422.16   
8               0.0               0.0              7.15        378.09   

   loc_og_mou_7  loc_og_mou_8  std_og_t2t_mou_6  std_og_t2t_mou_7  \
0           NaN          0.00               NaN               NaN   
7        533.91        255.79              4.30             23.29   
8        288.18         63.04            116.56            133.43   

   std_og_t2t_mou_8  std_og_t2m_mou_6  std_og_t2m_mou_7  std_og_t2m_mou_8  \
0              0.00               NaN               NaN              0.00   
7             12.01             49.89             31.76             49.14   
8             22.58             13.69             10.04             75.69   

   std_og_t2f_mou_6  std_og_t2f_mou_7  std_og_t2f_mou_8  std_og_t2c_mou_6  \
0               NaN               NaN              0.00               NaN   
7              6.66             20.08             16.68               0.0   
8              0.00              0.00              0.00               0.0   

   std_og_t2c_mou_7  std_og_t2c_mou_8  std_og_mou_6  std_og_mou_7  \
0               NaN               0.0           NaN           NaN   
7               0.0               0.0         60.86         75.14   
8               0.0               0.0        130.26        143.48   

   std_og_mou_8  isd_og_mou_6  isd_og_mou_7  isd_og_mou_8  spl_og_mou_6  \
0          0.00           NaN           NaN          0.00           NaN   
7         77.84           0.0          0.18         10.01           4.5   
8         98.28           0.0          0.00          0.00           0.0   

   spl_og_mou_7  spl_og_mou_8  og_others_6  og_others_7  og_others_8  \
0           NaN   

# EDA and Data Cleaning

In [31]:
# Getting the columns with only one unique value
cols_to_remove = []
for i in list(prime):
    if len(prime[i].value_counts()) == 1:
        cols_to_remove.append(i)
print(cols_to_remove)

['circle_id', 'loc_og_t2o_mou', 'std_og_t2o_mou', 'loc_ic_t2o_mou', 'last_date_of_month_6', 'last_date_of_month_7', 'last_date_of_month_8', 'std_og_t2c_mou_6', 'std_og_t2c_mou_7', 'std_og_t2c_mou_8', 'std_ic_t2o_mou_6', 'std_ic_t2o_mou_7', 'std_ic_t2o_mou_8']


In [32]:
# Droping columns with no variance 
prime = prime.drop(cols_to_remove, axis= 1)
prime.shape

(30001, 156)

In [34]:
# Percentage of null values in different columns
col_list = list(prime)
null = []
for i in prime.isnull().sum()*100/prime.shape[0]:
    null.append(i)
data = {"columns": col_list, "% null": null}
df_null = pd.DataFrame(data)
df_null[df_null["% null"] != 0]

columns     % null
4                 onnet_mou_6   1.816606
5                 onnet_mou_7   1.789940
6                 onnet_mou_8   3.913203
7                offnet_mou_6   1.816606
8                offnet_mou_7   1.789940
9                offnet_mou_8   3.913203
10              roam_ic_mou_6   1.816606
11              roam_ic_mou_7   1.789940
12              roam_ic_mou_8   3.913203
13              roam_og_mou_6   1.816606
14              roam_og_mou_7   1.789940
15              roam_og_mou_8   3.913203
16           loc_og_t2t_mou_6   1.816606
17           loc_og_t2t_mou_7   1.789940
18           loc_og_t2t_mou_8   3.913203
19           loc_og_t2m_mou_6   1.816606
20           loc_og_t2m_mou_7   1.789940
21           loc_og_t2m_mou_8   3.913203
22           loc_og_t2f_mou_6   1.816606
23           loc_og_t2f_mou_7   1.789940
24           loc_og_t2f_mou_8   3.913203
25           loc_og_t2c_mou_6   1.816606
26           loc_og_t2c_mou_7   1.789940
27           loc_og_t2c_mou_8   3.913203
28               loc_og_mou_6   1.816606
29               loc_og_mou_7   1.789940
30               loc_og_mou_8   3.913203
31           std_og_t2t_mou_6   1.816606
32           std_og_t2t_mou_7   1.789940
33           std_og_t2t_mou_8   3.913203
34           std_og_t2m_mou_6   1.816606
35           std_og_t2m_mou_7   1.789940
36           std_og_t2m_mou_8   3.913203
37           std_og_t2f_mou_6   1.816606
38           std_og_t2f_mou_7   1.789940
39           std_og_t2f_mou_8   3.913203
40               std_og_mou_6   1.816606
41               std_og_mou_7   1.789940
42               std_og_mou_8   3.913203
43               isd_og_mou_6   1.816606
44               isd_og_mou_7   1.789940
45               isd_og_mou_8   3.913203
46               spl_og_mou_6   1.816606
47               spl_og_mou_7   1.789940
48               spl_og_mou_8   3.913203
49                og_others_6   1.816606
50                og_others_7   1.789940
51                og_others_8   3.913203
55           loc_ic_t2t_mou_6   1.816606
56           loc_ic_t2t_mou_7   1.789940
57           loc_ic_t2t_mou_8   3.913203
58           loc_ic_t2m_mou_6   1.816606
59           loc_ic_t2m_mou_7   1.789940
60           loc_ic_t2m_mou_8   3.913203
61           loc_ic_t2f_mou_6   1.816606
62           loc_ic_t2f_mou_7   1.789940
63           loc_ic_t2f_mou_8   3.913203
64               loc_ic_mou_6   1.816606
65               loc_ic_mou_7   1.789940
66               loc_ic_mou_8   3.913203
67           std_ic_t2t_mou_6   1.816606
68           std_ic_t2t_mou_7   1.789940
69           std_ic_t2t_mou_8   3.913203
70           std_ic_t2m_mou_6   1.816606
71           std_ic_t2m_mou_7   1.789940
72           std_ic_t2m_mou_8   3.913203
73           std_ic_t2f_mou_6   1.816606
74           std_ic_t2f_mou_7   1.789940
75           std_ic_t2f_mou_8   3.913203
76               std_ic_mou_6   1.816606
77               std_ic_mou_7   1.789940
78               std_ic_mou_8   3.913203
82               spl_ic_mou_6   1.816606
83               spl_ic_mou_7   1.789940
84               spl_ic_mou_8   3.913203
85               isd_ic_mou_6   1.816606
86               isd_ic_mou_7   1.789940
87               isd_ic_mou_8   3.913203
88                ic_others_6   1.816606
89                ic_others_7   1.789940
90                ic_others_8   3.913203
100       date_of_last_rech_6   0.243325
101       date_of_last_rech_7   0.326656
102       date_of_last_rech_8   1.943269
106  date_of_last_rech_data_6  44.148528
107  date_of_last_rech_data_7  43.145228
108  date_of_last_rech_data_8  46.825106
109           max_rech_data_6  44.148528
110           max_rech_data_7  43.145228
111           max_rech_data_8  46.825106
112           count_rech_2g_6  44.148528
113           count_rech_2g_7  43.145228
114           count_rech_2g_8  46.825106
115           count_rech_3g_6  44.148528
116           count_rech_3g_7  43.145228
117           count_rech_3g_8  46.825106
124                 arpu_3g_6  44.148528
125 

In [35]:
# Let's check the columns with very large number (>= 40%) of null values
col_list = list(prime)
null = []
for i in prime.isnull().sum()*100/prime.shape[0]:
    null.append(i)
data = {"columns": col_list, "% null": null}
df_null = pd.DataFrame(data)
df_null[df_null["% null"] >= 40].shape

(25, 2)

So there are 25 columns with very high null values

In [38]:
# Checking above 25 columns 
correlation = []

lst = list(df_null[df_null["% null"] >= 40]["columns"])

NonNumCols = []
NumCols = []

for col in list(prime[lst]):
    if prime[lst][col].dtypes == "object":
        NonNumCols.append(col)  
    else:
        NumCols.append(col)
        
for i in NumCols:
    corr = round(list(prime[["churn", i]].corr()[i])[0], 3)
    correlation.append(corr)

print(correlation)

[0.002, -0.008, -0.027, 0.018, -0.003, -0.04, 0.034, 0.02, -0.013, 0.026, 0.007, -0.017, 0.031, 0.012, -0.014, 0.016, 0.003, -0.002, -0.011, -0.013, -0.079, -0.034]


**Observations**: Almost there is no correlation between these 25 columns and churn

In [39]:
# We can see that all the numerical columns with very high (>= 40%) null values are not of much importance for our model building.So, we can drop these columns
prime = prime.drop(NumCols, axis = 1)
prime.shape

(30001, 134)

In [40]:
# Imputing Categorical missing values with the mode
for i in NonNumCols:
    prime[i].fillna(prime[i].mode()[0], inplace=True)

In [41]:
# List of columns related to 'date'
date_cols = [x for x in list(prime) if "date" in x]
print(date_cols)

['date_of_last_rech_6', 'date_of_last_rech_7', 'date_of_last_rech_8', 'date_of_last_rech_data_6', 'date_of_last_rech_data_7', 'date_of_last_rech_data_8']


In [42]:
# Extracting day from the date
for i in date_cols:
    prime[i] = pd.to_datetime(prime[i]).dt.day

In [43]:
prime.head(3)

mobile_number    arpu_6    arpu_7    arpu_8  onnet_mou_6  onnet_mou_7  \
0     7000842753   197.385   214.816   213.803          NaN          NaN   
7     7000701601  1069.180  1349.850  3171.480        57.84        54.68   
8     7001524846   378.721   492.223   137.362       413.69       351.03   

   onnet_mou_8  offnet_mou_6  offnet_mou_7  offnet_mou_8  roam_ic_mou_6  \
0         0.00           NaN           NaN          0.00            NaN   
7        52.29        453.43        567.16        325.91          16.23   
8        35.08         94.66         80.63        136.48           0.00   

   roam_ic_mou_7  roam_ic_mou_8  roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  \
0            NaN           0.00            NaN            NaN           0.00   
7          33.49          31.64          23.74          12.59          38.06   
8           0.00           0.00           0.00           0.00           0.00   

   loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  loc_og_t2m_mou_6  \
0               NaN               NaN              0.00               NaN   
7             51.39             31.38             40.28            308.63   
8            297.13            217.59             12.49             80.96   

   loc_og_t2m_mou_7  loc_og_t2m_mou_8  loc_og_t2f_mou_6  loc_og_t2f_mou_7  \
0               NaN              0.00               NaN               NaN   
7            447.38            162.28             62.13             55.14   
8             70.58             50.54              0.00              0.00   

   loc_og_t2f_mou_8  loc_og_t2c_mou_6  loc_og_t2c_mou_7  loc_og_t2c_mou_8  \
0              0.00               NaN               NaN              0.00   
7             53.23               0.0               0.0              0.00   
8              0.00               0.0               0.0              7.15   

   loc_og_mou_6  loc_og_mou_7  loc_og_mou_8  std_og_t2t_mou_6  \
0           NaN           NaN          0.00               NaN   
7        422.16        533.91        255.79              4.30   
8        378.09        288.18         63.04            116.56   

   std_og_t2t_mou_7  std_og_t2t_mou_8  std_og_t2m_mou_6  std_og_t2m_mou_7  \
0               NaN              0.00               NaN               NaN   
7             23.29             12.01             49.89             31.76   
8            133.43             22.58             13.69             10.04   

   std_og_t2m_mou_8  std_og_t2f_mou_6  std_og_t2f_mou_7  std_og_t2f_mou_8  \
0              0.00               NaN               NaN              0.00   
7             49.14              6.66             20.08             16.68   
8             75.69              0.00              0.00              0.00   

   std_og_mou_6  std_og_mou_7  std_og_mou_8  isd_og_mou_6  isd_og_mou_7  \
0           NaN           NaN          0.00           NaN           NaN   
7         60.86         75.14         77.84           0.0          0.18   
8        130.26        143.48         98.28           0.0          0.00   

   isd_og_mou_8  spl_og_mou_6  spl_og_mou_7  spl_og_mou_8  og_others_6  \
0          0.00           NaN           NaN          0.00          NaN   
7         10.01           4.5           0.0          6.50          0.0   
8          0.00           0.0           0.0         10.23          0.0   

   og_others_7  og_others_8  total_og_mou_6  total_og_mou_7  total_og_mou_8  \
0          NaN          0.0            0.00            0.00            0.00   
7          0.0          0.0          487.53          609.24          350.16   
8          0.0          0.0          508.36          431.66          171.56   

   loc_ic_t2t_mou_6  loc_ic_t2t_mou_7  loc_ic_t2t_mou_8  loc_ic_t2m_mou_6  \
0               NaN               NaN              0.16               NaN   
7             58.14             32.26             27.31            217.56   
8             23.84              9.84              0.31             57.58   

   loc_ic_t2m_mou_7  loc_ic_t2m_mou_8  loc_ic_

In [44]:
# We still have few null vlaues which we will drop (dropping the rows with null values present)
for i in prime.columns:
    if prime[i].isnull().sum()>0:
        prime = prime[~np.isnan(prime[i])]

prime.shape

(28163, 134)

In [47]:
# Checking for null values again
col_list = list(prime)
null = []
for i in prime.isnull().sum()*100/prime.shape[0]:
    null.append(i)
data = {"columns": col_list, "% null": null}
df_null = pd.DataFrame(data)
df_null[df_null["% null"] != 0]

Empty DataFrame
Columns: [columns, % null]
Index: []

**Observation:**
The data now contain no null values and we can proceed ahead.